### ISSUE TITLE
Olingo4 olingo4endpointPropertyNames are missing queryParams for read method

### ISSUE THREAD

**Issue by varoliver** [apache/camel-quarkus#4654]
Hello,

we are migrating from camel 2.23 to camel quarkus (2.13 redhat).
Following code worked with 2.23 and appended the query parameters from the map which was generated by the method expression
```
.setHeader("CamelOlingo4.queryParams", method(SomeQueryBuilder.class, "buildQuery"))
.enrich("olingo4://read/SomeEntities", AggregationStrategies.flexible().storeInBody())
```
With camel quarkus the set
org.apache.camel.component.olingo4.Olingo4Endpoint#olingo4endpointPropertyNames
does not contain the value "queryParams". This results in calling toString() of the map in 
org.apache.camel.component.olingo4.Olingo4Endpoint#parseQueryParams
putting the value {key1=value1} in the local queryParams variable resulting in an invalid URL.
		
I think the issue may be that queryParams in Olingo4AppEndpointConfiguration is null and so
the key for queryParams does not get set in the properties map in:
org.apache.camel.support.component.ApiMethodPropertiesHelper#getConfigurationProperties

Please let us now if we are missing something else...
Kind regards
Oliver

**Comment by zhfeng** [comment]
It looks like a bug in camel-olingo4? can you raise an issue on https://issues.apache.org/jira/projects/CAMEL/

**Comment by jamesnetherton** [comment]
I created a ticket. Hopefully it captures the issue:

https://issues.apache.org/jira/browse/CAMEL-19150

**Comment by varoliver** [comment]
Thanks!

**Comment by zhfeng** [comment]
Thanks @jamesnetherton and the first idea in my mind is to add 
```
olingo4endpointPropertyNames.add(QUERY_PARAMS_PROPERTY);
```
in https://github.com/apache/camel/blob/main/components/camel-olingo4/camel-olingo4-component/src/main/java/org/apache/camel/component/olingo4/Olingo4Endpoint.java#L168

There is a generated API integration test `Olingo4AppIT` but all of method are marked with `@Disabled`. So I'm not sure if there is any test for these cases.

**Comment by varoliver** [comment]
@zhfeng thx, yes this would fix it, but would provide the property to every (http) METHOD.
I think it was not added in the first place because it is only relevant for READ and UREAD

The Olingo4AppEndpointConfigurationConfigurer does not copy the map because it is null in 
Olingo4AppEndpointConfiguration

Regards

**Comment by zhfeng** [comment]
`Olingo4AppEndpointConfigurationConfigurer` and `Olingo4AppEndpointConfiguration` are both generated by `camel-api-component-maven-plugin` automatically.

> The Olingo4AppEndpointConfigurationConfigurer does not copy the map because it is null in
Olingo4AppEndpointConfiguration

I'm not sure what are you meaning here. Can you show the codes for `copy the map`?

**Comment by varoliver** [comment]
private java.util.Map<String, String> queryParams;
in Olingo4AppEndpointConfiguration is null

so Olingo4AppEndpointConfigurationConfigurer returns in getOptionType
null for QueryParams when called from
org/apache/camel/support/component/ApiMethodPropertiesHelper.java:116
So queryparams does not get set into the map:
org.apache.camel.support.component.AbstractApiEndpoint#endpointProperties

if it would have been set it would get populated into the Map
olingo4endpointPropertyNames in
org/apache/camel/component/olingo4/Olingo4Endpoint.java:170

**Comment by zhfeng** [comment]
Thanks @varoliver and it seems that 
https://github.com/apache/camel/blame/main/components/camel-olingo4/camel-olingo4-component/src/main/java/org/apache/camel/component/olingo4/Olingo4Endpoint.java#L169

```
olingo4endpointPropertyNames = new HashSet<>(getEndpointPropertyNames());
```
which contains only nonNull propreties. But in camel-2.23.x, it uses
```
endpointPropertyNames = new HashSet<>(getPropertiesHelper().getValidEndpointProperties(configuration));
```
which contains all of the properties. I think this could explian it works in 2.23.x not fail in 3.x

Since these codes were changed by @davsclaus 3 years ago by https://github.com/apache/camel/commit/eba38a52355da86e22e2c985b336267d62df22b2

I wonder if it should change to use `getValidEndpointProperties` here.

**Comment by varoliver** [comment]
yes, mainly because the fields are propagated by reflection in 
`org.apache.camel.util.component.ApiMethodPropertiesHelper#getValidEndpointProperties`
independent of the value of the variable.
Unfortunate that there are no tests which covered this.
Small setback for our migration because we are using "CamelOlingo4.queryParams" quite a lot. We will have to change it to setting the parameters directly in the header like:
`exchange.getIn().setHeader("CamelOlingo4.$filter", "someValue");`

**Comment by zhfeng** [comment]
I open a PR to fix it in upstream camel
- https://github.com/apache/camel/pull/9570

@varoliver can you take a look?

**Comment by varoliver** [comment]
@zhfeng looks good. What i am wondering is what will happen for another METHOD like PATCH or DELETE in case queryParam is set in exchange header - would the params also be appended to the url now?

I read the fix will be backported also to 3.x.  This would be good for us.
Current Camel 3.18. version is 3.18.5. 
**So the fix should be available in 3.18.6?** If so - when do u think it will be released?
Red Hat build of Camel Extensions 2.13.2 uses Camel 3.18.3.
Will there be an build of Camel Extensions for Quarkus 2.13.3 with 3.18.6 (containing the fix)?
If there will be a build but no official Red Hat build would it be safe to reference the olingo4 module v2.13.3 in our pom but depend on Red Hat Camel Extension 2.13.2 bom?

THX

**Comment by zhfeng** [comment]
@varoliver I just go throught the codes and it looks like other method which does not have `queryParams` in the args will not get it from the properties. See https://github.com/apache/camel/blob/main/core/camel-support/src/main/java/org/apache/camel/support/component/ApiMethodHelper.java#L478-L484

So I belive it should not append to the url even if `queryParams` is set in the headers. 

Regards for the Red Hat build of Camel Extensions for Quarkus, please contact them for supporting.

**Comment by zhfeng** [comment]
The camel-quarkus `2.13.3` release should contains this fix in camel `3.18.6`. So I'd like to close this issue. @varoliver feel free to test with `2.13.3` and reopen it if there is any problem. Thanks!

**Comment by ppalaga** [comment]
Could please somebody set the right milestone where this was fixed?
